_posts/2018-12-27-reactive-programming-in-java.markdown

```diff
@@ -0,0 +1,110 @@
+
+### Introduction to RxJava
+There are different implementation on Java for reactive programming, like:
+- Java 9 Flow API
+- RxJava
+- Reactor 
+
+Here I would like to take RxJava as an example. 
+To use RxJava in our Maven project, we’ll need to add the following dependency to our pom.xml:
+```
+<dependency>
+    <groupId>io.reactivex</groupId>
+    <artifactId>rxjava</artifactId>
+    <version>${rx.java.version}</version>
+</dependency>
+```
+There are two key types to understand when working with RxJava:
+
+- Observable represents any object that can get data from a data source and whose state may be of interest in a way that other objects may register an interest
+- An observer is any object that wishes to be notified when the state of another object changes
+An observer subscribes to an Observable sequence. The sequence sends items to the observer one at a time.
+
+The observer handles each one before processing the next one. If many events come in asynchronously, they must be stored in a queue or dropped.
+
+In Rx, an observer will never be called with an item out of order or called before the callback has returned for the previous item.
+Create Observable like below:
+```
+Observable<String> observable = Observable.just("Hello");
+observable.subscribe(s -> result = s);
+  
+assertTrue(result.equals("Hello"));
+```
+The example on three different methods on the observer interface: OnNext, OnError, and OnCompleted
+```
+String[] letters = {"a", "b", "c", "d", "e", "f", "g"};
+Observable<String> observable = Observable.from(letters);
+observable.subscribe(
+  i -> result += i,  //OnNext
+  Throwable::printStackTrace, //OnError
+  () -> result += "_Completed" //OnCompleted
+);
+assertTrue(result.equals("abcdefg_Completed"));
+```
+The samle for itertor:
+```
+List<String> words = Arrays.asList(
+ "the",
+ "quick",
+ "brown",
+ "fox",
+ "jumped",
+ "over",
+ "the",
+ "lazy",
+ "dogs"
+);
+
+Observable.fromIterable(words)
+ .flatMap(word -> Observable.fromArray(word.split("")))
+ .distinct()
+ .sorted()
+ .zipWith(Observable.range(1, Integer.MAX_VALUE),
+   (string, count) -> String.format("%2d. %s", count, string))
+ .subscribe(System.out::println);
+```
+with the result:
+```
+ 1. a
+ 2. b
+ 3. c
+ 4. d
+ 5. e
+ 6. f
+ 7. g
+ 8. h
+ 9. i
+10. j
+11. k
+12. l
+13. m
+14. n
+15. o
+16. p
+17. q
+18. r
+19. s
+20. t
+21. u
+22. v
+23. w
+24. x
+25. y
+26. z
+```
+### Conclusion
+We've talked about the concept of reactive programming, the benefit, how can we use it and different implementation and take RxJava as the example. Basically, the reactive programming is a concept and methodology, different people or group might have different view or interpretation on it but one simple word on the purpose is to make the application more responsive to provide the high performance to our users
```

Comment:
@u9520107 resolved all the grammar issues, please check again. Thanks a lot for your suggestion.